### PR TITLE
[full-ci] experimental tags backport

### PR DESCRIPTION
**Pull request by fschade**
## Description
Backport of experimental tags implementation, instead of using rest to access the tag endpoints this implementation is using graph to add these.

## Related Issue

- Part of https://github.com/owncloud/ocis/issues/5184
- Needs https://github.com/owncloud/libre-graph-api/pull/62

### CODE REVIEW

File: services/graph/pkg/service/v0/tags.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	"github.com/cs3org/reva/v2/pkg/storagespace"
+	revaTags "github.com/cs3org/reva/v2/pkg/tags"
+	"github.com/go-chi/render"
+	libregraph "github.com/owncloud/libre-graph-api-go"
+	searchsvc "github.com/owncloud/ocis/v2/protogen/gen/ocis/services/search/v0"
+	"github.com/owncloud/ocis/v2/services/graph/pkg/service/v0/errorcode"
+	"go-micro.dev/v4/metadata"
+)
+
+func (g Graph) GetTags(w http.ResponseWriter, r *http.Request) {
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:22]: Exported func `GetTags` should have comment or be unexported

File: services/graph/pkg/service/v0/tags.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	tagCollection := libregraph.NewCollectionOfTags()
+	tagCollection.Value = tagList.AsSlice()
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, tagCollection)
+}
+
+func (g Graph) AssignTags(w http.ResponseWriter, r *http.Request) {
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:50]: Exported func `AssignTags` should have comment or be unexported

File: services/graph/pkg/service/v0/tags.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+			SpaceOwner: sres.Info.Owner,
+			Executant:  revaCtx.ContextMustGetUser(r.Context()).Id,
+		}
+		if err := events.Publish(g.eventsPublisher, ev); err != nil {
+			g.logger.Error().Err(err).Msg("Failed to publish TagsAdded event")
+		}
+	}
+}
+
+func (g Graph) UnassignTags(w http.ResponseWriter, r *http.Request) {
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:139]: Exported func `UnassignTags` should have comment or be unexported

File: services/graph/pkg/service/v0/tags.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+import (
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	revaCtx "github.com/cs3org/reva/v2/pkg/ctx"
+	revactx "github.com/cs3org/reva/v2/pkg/ctx"
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:11]: imported twice

File: services/graph/pkg/service/v0/tags.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	"github.com/cs3org/reva/v2/pkg/events"
+	"github.com/cs3org/reva/v2/pkg/storagespace"
+	revaTags "github.com/cs3org/reva/v2/pkg/tags"
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:14]: Curious: Why alias this with `revaTags`? Whats wrong with `tags`?

File: services/graph/pkg/service/v0/tags.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+		return
+	}
+
+	var currentTags string
+	if m := sres.GetInfo().GetArbitraryMetadata().GetMetadata(); m != nil {
+		currentTags = m["tags"]
+	}
+
+	allTags := revaTags.FromList(currentTags)
+	newTags := strings.Join(assignment.Tags, ",")
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:103]: Why is it needed to add a `,` to the tags string? Bug in `reva/tags` pkg?

File: services/graph/pkg/service/v0/tags.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	if !pm.InitiateFileUpload && !pm.CreateContainer {
+		g.logger.Info().Msg("no permission to create a tag")
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	var currentTags string
+	if m := sres.GetInfo().GetArbitraryMetadata().GetMetadata(); m != nil {
+		currentTags = m["tags"]
+	}
```
Comment (kobergj) [inline services/graph/pkg/service/v0/tags.go:189]: Afaics this code part is 1:1 the same as in `AssignTags`. Can we put this in a func that both are using?

**Review by kobergj — CHANGES_REQUESTED** [review]
Some questions...

File: services/graph/pkg/service/v0/tags.go (hunk 5 above)
Comment (fschade) [reply services/graph/pkg/service/v0/tags.go:14]: assignment.Tags is a slice of tags ([]string), tags package does not have a `fromSlice` func at the moment.

File: services/graph/pkg/service/v0/tags.go (hunk 6 above)
Comment (fschade) [reply services/graph/pkg/service/v0/tags.go:103]: assignment.Tags is a slice of tags ([]string), tags package does not have a fromSlice func at the moment.

Comment (kobergj) [reply services/graph/pkg/service/v0/tags.go:104]: Hehe. OK. So it IS a bug in `reva/tags` pkg :smile:

**Review by kobergj — APPROVED** [review]

**Pull request merged**